src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -69,15 +70,141 @@ internal sealed partial class PkcsPalOpenSsl : PkcsPal
                     
                     if (unprotectedAttributes.Count != 0)
                     {
-                        // TODO(3334): Add support for unprotected attributes here.
-                        throw new NotImplementedException();
+                        encryptedMessage = AddAttributesToEncoding(encryptedMessage, unprotectedAttributes);
                     }
 
                     return encryptedMessage;
                 }
             }
         }
 
+        private byte[] AddAttributesToEncoding(byte[] encryptedMessage, CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            DerSequenceReader originalMessage = new DerSequenceReader(encryptedMessage);
+
+            // EnvelopedCms has the following underlying structure:
+            //
+            // ContentInfo ::= SEQUENCE {
+            //      contentType ContentType,
+            //      content[0] EXPLICIT ANY DEFINED BY contentType }
+            //
+            // ContentType ::= OBJECT IDENTIFIER
+            //
+            // According to RFC2630, the enveloped type we have support for has the following structure
+            // EnvelopedData ::= SEQUENCE {
+            //     version CMSVersion,
+            //     originatorInfo[0] IMPLICIT OriginatorInfo OPTIONAL,
+            //     recipientInfos RecipientInfos,
+            //     encryptedContentInfo EncryptedContentInfo,
+            //     unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL }
+            //
+            // As well as it remarks that "If unprotectedAttrs is present, then version shall be 2."
+
+            byte[][] contentType = originalMessage.ReadAndSplitNextEncodedValue();
+            DerSequenceReader envelopedDataReader = originalMessage.ReadExplicitContextSequence();
+
+            int versionNumber = envelopedDataReader.ReadInteger();
+
+            if (versionNumber > 4)
+                throw new PlatformNotSupportedException(SR.Cryptography_Cms_VersionNumberPlatformNotSupported);
+
+            if (versionNumber == 0)
+                versionNumber = 2;
+
+            byte[] version = ConcatenateArrays(DerEncoder.SegmentedEncodeUnsignedInteger(new byte[] { (byte)versionNumber }));
+
+            // We don't need to parse the OriginatorInfo, RecipientInfos, or EncryptedContentInfo, so
+            // we can just extract it to concatenate it.
+            byte[] remainingDataOfSequence = envelopedDataReader.ReadRemainingData();
+
+            byte[][] envelopedData = DerEncoder.ConstructSegmentedExplicitSequenceFromPayload(
+                0 /* Context number */,
+                version,
+                remainingDataOfSequence,
+                EncodeUnprotectedAttributes(unprotectedAttributes));
+
+            return DerEncoder.ConstructSequence(
+                contentType,
+                envelopedData);
+        }
+
+        private static byte[] ConcatenateArrays(params byte[][] data)
```

Comment:
This isn't being used paramsally, so you can get rid of the params keyword.  (The only caller is passing a byte[][])